File: codes/group-anagrams.py
from typing import List


class Solution:
    @staticmethod
    def groupAnagrams(strs: List[str]) -> List[List[str]]:
        anagrams = {}
        for word in strs:
            sorted_str = "".join(sorted(word))
            if anagrams.get(sorted_str):
                anagrams[sorted_str] = [*anagrams[sorted_str], word]
            else:
                anagrams[sorted_str] = [word]

        return [*anagrams.values()]


print(
    Solution.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
)  # [["bat"],["nat","tan"],["ate","eat","tea"]]

File: codes/arrays_and_hashing/group-anagrams.rb
# frozen_string_literal: true

# @param {String[]} strs
# @return {String[][]}
def group_anagrams(strs)
  anagrams = {}

  strs.each do |word|
    sorted_word = word.chars.sort.join

    anagrams[sorted_word] = [*anagrams[sorted_word], word]
  end

  anagrams.values
end

p group_anagrams %w[eat tea tan ate nat bat] # [["eat", "tea", "ate"], ["tan", "nat"], ["bat"]]
